python/ray/tune/execution/ray_trial_executor.py

```diff
@@ -498,13 +499,21 @@ def _stop_trial(
         Args:
             error: Whether to mark this trial as terminated in error.
             exc: Optional exception.
+            should_increment_num_failures_counter: Whether to increment per trial
+                failure counter.
+                This is False when the trial is using another restore retry counter.
         """
         self.set_status(trial, Trial.ERROR if error or exc else Trial.TERMINATED)
         self._trial_just_finished = True
         trial.set_location(_Location())
 
         try:
-            trial.write_error_log(exc=exc)
+            trial.write_error_log(
+                exc=exc,
+                should_increment_num_failures_counter=(
+                    should_increment_num_failures_counter
+                ),
+            )
```

Comment:
This leads to a pretty strong coupling of retry-logic between trial runner and trial.

I think ideally we would move this whole logic just into the trial, which keeps track of it's errors and retry-behavior.

For this, the trial would have to identify the cause of the error (e.g. actor dead, training fn failure, restore failure) and update it's own state accordingly and react in `should_recover()`. Then the trial runner does not have to do any special casing